File: ui/packages/ui/src/search/engine/query.ts
import { searchFieldHit } from "./field";
import { fieldWeight, scoreKind } from "./score";
import { tokenizeSearchQuery } from "./token";
import type { SearchCombine, SearchDocument, SearchHit, SearchMatch, SearchOptions } from "./types";

function matchDocument<T>(
  doc: SearchDocument<T>,
  tokens: readonly string[],
  index: number,
  combine: SearchCombine,
): SearchMatch<T> | null {
  if (tokens.length === 0) {
    return { id: doc.id, item: doc.item, score: 0, hits: [], group: doc.group, index };
  }
  const hits: SearchHit[] = [];
  let score = 0;
  let covered = 0;
  for (const token of tokens) {
    let best = 0;
    let matched = false;
    for (const field of doc.fields) {
      const hit = searchFieldHit(field.text, token);
      if (!hit) continue;
      matched = true;
      hits.push({ ...hit, field: field.key });
      best = Math.max(best, scoreKind(hit.kind, fieldWeight(field)));
    }
    if (matched) {
      covered += 1;
      score += best;
    }
  }
  if (covered === 0) return null;
  if (combine === "and" && covered < tokens.length) return null;
  return { id: doc.id, item: doc.item, score, hits, group: doc.group, index };
}

function emptyMatches<T>(docs: readonly SearchDocument<T>[]): SearchMatch<T>[] {
  return docs.map((doc, index) => ({
    id: doc.id,
    item: doc.item,
    score: 0,
    hits: [],
    group: doc.group,
    index,
  }));
}

export function searchDocuments<T>(
  docs: readonly SearchDocument<T>[],
  query: string,
  options: SearchOptions = {},
): SearchMatch<T>[] {
  const tokens = tokenizeSearchQuery(query);
  const empty = options.empty ?? "pass";
  if (tokens.length === 0) return empty === "none" ? [] : emptyMatches(docs);
  const combine = options.combine ?? "and";
  const out: SearchMatch<T>[] = [];
  docs.forEach((doc, index) => {
    const match = matchDocument(doc, tokens, index, combine);
    if (match) out.push(match);
  });
  out.sort((a, b) => b.score - a.score || a.index - b.index);
  return out;
}

File: core/yohu-search/src/query.rs
//! 文档表检索。默认 AND + 空查询原序通过。

use crate::field::search_field_hit;
use crate::score::{field_weight, kind_score};
use crate::token::tokenize_search_query;
use crate::types::{SearchCombine, SearchDocument, SearchEmpty, SearchHit, SearchMatch, SearchOptions};

fn match_document(
    doc: &SearchDocument,
    tokens: &[String],
    index: usize,
    combine: SearchCombine,
) -> Option<SearchMatch> {
    if tokens.is_empty() {
        return Some(SearchMatch {
            id: doc.id.clone(),
            score: 0,
            hits: Vec::new(),
            group: doc.group.clone(),
            index,
        });
    }
    let mut hits = Vec::new();
    let mut score = 0;
    let mut covered = 0usize;
    for token in tokens {
        let mut best = 0;
        let mut matched = false;
        for field in &doc.fields {
            if let Some((start, end, kind)) = search_field_hit(&field.text, token) {
                matched = true;
                hits.push(SearchHit {
                    field: field.key.clone(),
                    start,
                    end,
                    kind,
                });
                best = best.max(kind_score(kind) * field_weight(field));
            }
        }
        if matched {
            covered += 1;
            score += best;
        }
    }
    if covered == 0 {
        return None;
    }
    if combine == SearchCombine::And && covered < tokens.len() {
        return None;
    }
    Some(SearchMatch {
        id: doc.id.clone(),
        score,
        hits,
        group: doc.group.clone(),
        index,
    })
}

fn empty_matches(docs: &[SearchDocument]) -> Vec<SearchMatch> {
    docs.iter()
        .enumerate()
        .map(|(index, doc)| SearchMatch {
            id: doc.id.clone(),
            score: 0,
            hits: Vec::new(),
            group: doc.group.clone(),
            index,
        })
        .collect()
}

pub fn search_documents(
    docs: &[SearchDocument],
    query: &str,
    options: SearchOptions,
) -> Vec<SearchMatch> {
    let tokens = tokenize_search_query(query);
    if tokens.is_empty() {
        return match options.empty {
            SearchEmpty::None => Vec::new(),
            SearchEmpty::Pass => empty_matches(docs),
        };
    }
    let mut out: Vec<SearchMatch> = docs
        .iter()
        .enumerate()
        .filter_map(|(index, doc)| match_document(doc, &tokens, index, options.combine))
        .collect();
    out.sort_by(|a, b| b.score.cmp(&a.score).then(a.index.cmp(&b.index)));
    out
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde::Deserialize;

    #[derive(Deserialize)]
    struct Case {
        query: String,
        #[serde(default)]
        options: SearchOptions,
        ids: Vec<String>,
        #[serde(default)]
        score0: Option<i32>,
    }

    #[derive(Deserialize)]
    struct Fixture {
        docs: Vec<SearchDocument>,
        search: Vec<Case>,
    }

    #[test]
    fn search_shared_fixture() {
        let data: Fixture =
            serde_json::from_str(include_str!("../testdata/search.json")).expect("fixture");
        for (i, case) in data.search.iter().enumerate() {
            let matches = search_documents(&data.docs, &case.query, case.options);
            let ids: Vec<&str> = matches.iter().map(|m| m.id.as_str()).collect();
            assert_eq!(
                ids,
                case.ids.iter().map(String::as_str).collect::<Vec<_>>(),
                "search {i}"
            );
            if let Some(score) = case.score0 {
                assert_eq!(matches.first().map(|m| m.score), Some(score), "score {i}");
            }
        }
    }
}
